WRQ-5674: Support to change video resolution on video player

mmyelyn opened the pull request:
### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
This feature was created by a request to enable video resolution change in our video player. 
So We've created guidelines to help support this functionality.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
Add video resolution change support using hls.js to the pattern-video-player-custom sample

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

### Links
[//]: # (Related issues, references)
WRQ-5674

### Comments
Enact-DCO-1.0-Signed-off-by: Hyelyn Kim ([myelyn.kim@lge.com](mailto:sc.baek@lge.com))

vJIYEv committed:
Create pattern-video-player-custom app
Enact-DCO-1.0-Signed-off-by: Jiye Kim (jiye.kim@lge.com)

vJIYEv committed:
update package.json dependencies: hls.js

mmyelyn committed:
modify sample to support changing video resolution

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/enactjs/samples?pullRequest=218) <br/>All committers have signed the CLA.

mmyelyn committed:
Merge branch 'develop' into feature/WRQ-5674

mmyelyn committed:
Delete unnecessary code added during merge

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn committed:
Add new line

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn committed:
Merge branch 'develop' into feature/WRQ-5674

juwonjeong commented:
Please update README.md file

mmyelyn committed:
Applied review

- Fixed an issue where the resolution button did not appear when the panel was hidden
- Update README

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn committed:
Update README.md

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

juwonjeong commented on package.json:
```diff
@@ -22,4 +22,5 @@
-}
\ No newline at end of file
+}
+
```
It seems that an unnecessary line is added. Please check it.

mmyelyn replied:
Removed

juwonjeong commented on sandstone/pattern-video-player-custom/README.md:
```diff
@@ -16,6 +16,7 @@ Run `npm install` then `npm run serve` to have the app running on [http://localh
 The video data, including the URL and mime type, are stored in [videos.js](src/App/videos.js). For `m3u8` videos, the mime type should be `application/x-mpegURL`.
 To enable `m3u8` video playback, the application uses [hls.js](https://github.com/video-dev/hls.js), which is imported in [App.js](src/App/App.js).
 In [App.js](src/App/App.js), the application retrieves the video data from [videos.js](src/App/videos.js) and assigns the source and video element to the VideoPlayer based on the mime type.
+Functionally, you can change the resolution of the video in m3u8 video.
```
I think it would be better to update the descriptions as mentioned above. 
```suggestion
Functionally, you can change the resolution of the `m3u8` video.
```

mmyelyn replied:
I Updated.

juwonjeong commented on sandstone/pattern-video-player-custom/src/App/App.js:
```diff
@@ -62,6 +87,25 @@ const AppBase = ({className, videoId, ...rest}) => {
+						<div> Select Resolution </div>
+						<br />
+						<div>
+							{resolutions.map((resolution, index) => {
```
Please check the console warning.

mmyelyn replied:
I added key prop. I want to key value as `resolution.id` but this value was received always 0. So I used `resolution._attrs[0].RESOLUTION`. Alternatively, although it is not recommended, I can also use an index.

mmyelyn committed:
Applied review

- Added key prop for map statement
- Updated README

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn committed:
Removed unnecessary empty line

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

juwonjeong commented on sandstone/pattern-video-player-custom/package.json:
```diff
@@ -43,4 +43,4 @@
     "eslint-config-enact-proxy": "^1.0.7"
   }
-}
+}
```
Please revert this change

mmyelyn committed:
Added empty line

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn committed:
Merge branch 'develop' into feature/WRQ-5674

mmyelyn commented:
A conflict occurred because the README content previously included in develop was missing.
(Merged into develop in #222, but reverted in #221)
Since it is correct that the README content is being updated, I have added it again in the PR.

juwonjeong committed:
Update README.md

Enact-DCO-1.0-Signed-off-by: Juwon Jeong (juwon.jeong@lge.com)

juwonjeong commented:
> A conflict occurred because the README content previously included in develop was missing. (Merged into develop in #222, but reverted in #221) Since it is correct that the README content is being updated, I have added it again in the PR.

Okay, I also added a change that was dropped.

juwonjeong reviewed (approved):
LGTM

seunghoh reviewed (changes requested):
Please merge the README and js files when https://github.com/enactjs/samples/pull/219 merged.

mmyelyn committed:
Merge branch 'develop' into feature/WRQ-5674

seunghoh reviewed (changes requested):
1. The resolution selection panel stays open and wait , the video control is hidden. what happens then? 
2. I don't know which resolution we selected.
3. How can I go to the default? (no selection and automation)

mmyelyn committed:
Changed implementation using dropdown rather than panel with button

- Added auto selection
- Added implementation to know the selected resolution

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn commented:
> 1. The resolution selection panel stays open and wait , the video control is hidden. what happens then?
> 2. I don't know which resolution we selected.
> 3. How can I go to the default? (no selection and automation)

1. Thank you. Instead of using overlay, I changed the implementation to dropdown.
2. This was resolved with a change to the dropdown. You can see the selected resolution.
3. Added default value 'auto'. If not selected, the default value is set to auto.

seunghoh reviewed (changes requested):
Dropdown doesn't keep the value that I've selected. 
For example, select the resolution 640*480 -> close the panel -> open the panel again but dropdown selects Auto.

mmyelyn committed:
Fixed issue where it returns to auto when selecting the first index resolution

- Fixed edge case..

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn commented:
> Dropdown doesn't keep the value that I've selected. For example, select the resolution 640*480 -> close the panel -> open the panel again but dropdown selects Auto.

I found an edge case. When I selected the first item in the dropdown, I checked that the index was 0, so it became a falsy value. Fixed this.

seunghoh committed:
Fix cleanup code and update readme

Enact-DCO-1.0-Signed-off-by: Seungho Park <seunghoh.park@lge.com>

seunghoh commented:
Merged, but I slightly updated your PR. @mmyelyn
